### PR TITLE
Flexibility Analysis

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:

This PR adds a module for formulating flexibility analysis problems.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by michaelbynum** [comment]
How do I get the integration tests to run?

**Comment by andrewlee94** [comment]
@michaelbynum As you worked out, you add the CI:Integration tag to trigger the run. Alternatively, if the PR has two approvals they will be triggered on a new push.

**Comment by michaelbynum** [comment]
Thanks!

File: idaes/apps/flexibility_analysis/_check_relu_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#################################################################################
+"""
+This module is only used to check dependencies for unit tests.
+"""
+import unittest
+from pyomo.common.dependencies import attempt_import
+
+tensorflow, tensorflow_available = attempt_import("tensorflow")
+omlt, nump_available = attempt_import("omlt")
+if not tensorflow_available or not nump_available:
```
Comment (lbianchi-lbl) [inline idaes/apps/flexibility_analysis/_check_relu_dependencies.py:21]:
```suggestion
omlt, omlt_available = attempt_import("omlt")
if not tensorflow_available or not omlt_available:
```

File: idaes/apps/flexibility_analysis/_check_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+This module is only used to check dependencies for unit tests.
+"""
+import unittest
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/_check_dependencies.py:16]: Seeing as this is just for testing, should it be moved into the `tests` folder?

File: idaes/apps/flexibility_analysis/_check_relu_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+This module is only used to check dependencies for unit tests.
+"""
+import unittest
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/_check_relu_dependencies.py:16]: As above.

File: idaes/apps/flexibility_analysis/check_optimal.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+This module provides a function like Pyomo's assert_optimal_termination but that 
+works for both APPSI solver interfaces and non-appsi solver interfaces.
+"""
+import pyomo.environ as pe
+from pyomo.contrib import appsi
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/check_optimal.py:18]: A minor comment (so does not need to be fixed), but this seems like it would be better located in Pyomo (probably in `pyomo.contrib` with the `appsi` code seeing as it related to that).

File: idaes/apps/flexibility_analysis/decision_rules/dr_config.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/dr_config.py:34]: Is the purpose of this just to change the default values for the `implicit` and `implicit_domain` arguments (whilst still leaving them as arguments)?

File: idaes/apps/flexibility_analysis/decision_rules/linear_dr.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+        self.solver = self.declare(
+            "solver", ConfigValue(default=pe.SolverFactory("ipopt"))
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/linear_dr.py:54]: Should this also have a `doc` attribute, just in case someone tries to query the ConfigDict?

File: idaes/apps/flexibility_analysis/decision_rules/linear_dr.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            lin_vars = list(trainer.coefs.values())
+            lin_coefs.append(1)
+            lin_vars.append(trainer.const)
+            lin_coefs.append(-1)
+            lin_vars.append(trainer.out_est[ndx])
+            expr = LinearExpression(linear_coefs=lin_coefs, linear_vars=lin_vars)
+            trainer.est_cons[ndx] = (expr, 0)
+
+        results = config.solver.solve(trainer)
+        assert_optimal_termination(results)
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/linear_dr.py:116]: Should this have a try/except in order to provide a useful error message to the user, or will this be caught elsewhere?

File: idaes/apps/flexibility_analysis/decision_rules/linear_dr.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        A pyomo model containing the linear decision rule
+    """
+    n_inputs = len(input_vals)
+    n_outputs = len(output_vals)
+
+    res = pe.Block(concrete=True)
+    res.output_set = pe.Set(initialize=list(range(n_outputs)))
+    res.decision_rule = pe.Constraint(res.output_set)
+
+    for out_ndx, (output_var, out_samples) in enumerate(output_vals.items()):
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/linear_dr.py:88]: For later maintainers, some inline docs would be helpful. I can work out what each part is doing, but I don't know why it is doing it.

File: idaes/apps/flexibility_analysis/decision_rules/relu_dr_config.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            visibility=visibility,
+        )
+        self.n_layers: int = self.declare(
+            "n_layers", ConfigValue(domain=int, default=4)
+        )
+        self.n_nodes_per_layer: int = self.declare(
+            "n_nodes_per_layer", ConfigValue(domain=int, default=4)
+        )
+        self.tensorflow_seed: int = self.declare(
+            "tensorflow_seed", ConfigValue(domain=int, default=0)
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/relu_dr_config.py:72]: Is it a good idea to have a default value for the seed? From my (limited) understanding, setting the seed would mean the sequence is pre-determined. Whilst this is good for testing, I would assume this is not what you want for actual code (otherwise why use `random`).

File: idaes/apps/flexibility_analysis/decision_rules/relu_dr_config.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+        self.n_layers: int = self.declare(
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/relu_dr_config.py:65]: These might be obvious to people with more experience with neural networks, but it might be good to have some short doc strings for these arguments for less experiences users.

File: idaes/apps/flexibility_analysis/decision_rules/relu_dr.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        A pyomo model containing the linear decision rule
+    """
+    tf.random.set_seed(config.tensorflow_seed)
+    inputs = list(input_vals.keys())
+    outputs = list(output_vals.keys())
+    n_samples = len(input_vals[inputs[0]])
+
+    config: ReluDRConfig = config()
+    if config.batch_size > n_samples:
+        config.batch_size = n_samples
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/relu_dr.py:64]: It might be good to log a message for the user here to let them know they set a batch size greater than the number of samples and thus this value was changed.

File: idaes/apps/flexibility_analysis/decision_rules/relu_dr.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    nn = keras.Sequential()
+    nn.add(
+        keras.layers.Dense(
+            units=config.n_nodes_per_layer, input_dim=len(inputs), activation="relu"
+        )
+    )
+    for _ in range(config.n_layers - 1):
+        nn.add(keras.layers.Dense(config.n_nodes_per_layer, activation="relu"))
+    nn.add(keras.layers.Dense(len(outputs)))
+    if config.learning_rate is None:
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/relu_dr.py:105]: I assume this is needed as `keras` does not accept `None` as a value for `learning_rate` (i.e. if you provide the argument it must have a meaningful value). An inline comment might help to make sure no-one tries to "fix" this later.

File: idaes/apps/flexibility_analysis/decision_rules/relu_dr.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        opt = keras.optimizers.Adam()
+    else:
+        opt = keras.optimizers.Adam(learning_rate=config.learning_rate)
+    nn.compile(optimizer=opt, loss="mse")
+    history = nn.fit(
+        training_input,
+        training_output,
+        batch_size=config.batch_size,
+        epochs=config.epochs,
+        # verbose=0,
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/relu_dr.py:115]: Commented line.

File: idaes/apps/flexibility_analysis/decision_rules/tests/test_linear_dr.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+from idaes.apps.flexibility_analysis import _check_dependencies
```
Comment (andrewlee94) [inline idaes/apps/flexibility_analysis/decision_rules/tests/test_linear_dr.py:13]: Should there be tests for the relu_dr as well?

**Review by andrewlee94 — COMMENTED** [review]
Here are some comments on the first file (I'll try to get through the rest tomorrow).

One general comment; could you add yourself to the top-level CODEOWNERS file as the maintainer for this code (or a nominee).

File: requirements-dev.txt [hunk 15]
```diff
@@ -29,6 +30,7 @@ jsonschema
 jupyter_contrib_nbextensions
 snowballstemmer==1.2.1
 addheader>=0.2.2
+highspy
```
[the remark this answers is not in the document]
Comment (michaelbynum) [reply requirements-dev.txt:33]: SCIP would work, but it makes the tests way slower. I'm solving a bunch of LPs with changes only to parameters. The Pyomo interface to HiGHS is a persistent interface, and is much faster than the interface to SCIP for that type of workflow.

File: idaes/apps/flexibility_analysis/_check_dependencies.py (hunk 2 above)
Comment (michaelbynum) [reply idaes/apps/flexibility_analysis/_check_dependencies.py:16]: Great point. I'll move this.

File: idaes/apps/flexibility_analysis/check_optimal.py (hunk 4 above)
Comment (michaelbynum) [reply idaes/apps/flexibility_analysis/check_optimal.py:18]: Great point.

File: requirements-dev.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewlee94) [reply requirements-dev.txt:33]: We might want to see if it makes sense to do something like @lbianchi-lbl did for SCIP then so that it does not need to be a dependency in requirements-dev.txt. If I recall, we basically have a check that sees if SCIP is available and otherwise skips the tests. We then just make sure we have it installed on the test server (via GHA setup) so the test will run on the server.

File: idaes/apps/flexibility_analysis/decision_rules/dr_config.py (hunk 5 above)
Comment (michaelbynum) [reply idaes/apps/flexibility_analysis/decision_rules/dr_config.py:34]: I don't think I changed any defaults. This class is just meant to be a base class for other, related config classes. I guess it is not really necessary. I think I just created it for type hinting.

File: idaes/apps/flexibility_analysis/decision_rules/linear_dr.py (hunk 6 above)
Comment (michaelbynum) [reply idaes/apps/flexibility_analysis/decision_rules/linear_dr.py:54]: Absolutely. I'll add one.

File: requirements-dev.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (michaelbynum) [reply requirements-dev.txt:33]: Sounds good to me. As long as it gets installed in GHA so the tests run, I'm happy.

[the remark this answers is not in the document]
Comment (lbianchi-lbl) [reply requirements-dev.txt:33]: Not sure if it applies, but would the other way around work? i.e. is it possible to use HiGHS instead of SCIP for our existing tests for which SCIP is currently being used?

If so, it might make sense to consider having `highspy` as a dev dependency and just use it for testing for both cases (of course, this is purely on the side of streamlining dependencies and very much solver-domain-ignorant, so feel free to add why this wouldn't work).

[the remark this answers is not in the document]
Comment (michaelbynum) [reply requirements-dev.txt:33]: I actually need both. SCIP also solves nonlinear problems, while HiGHS does not...

[the remark this answers is not in the document]
Comment (lbianchi-lbl) [reply requirements-dev.txt:33]: Ah, that makes sense then, thanks for clarifying!

**Comment by ksbeattie** [comment]
@michaelbynum any update here?

**Comment by andrewlee94** [comment]
@michaelbynum Something you might want to think about for the docs here: I have received a number of questions this week about this PR and how it is distinguished from robust optimization and Pyros (and thus when you would want to use one versus the other). It would probably be good to have some discussion of this somewhere in the docs to guide new users to the best tool for what they want to do. On the plus side, there is already some interest in tools like this.

**Comment by ksbeattie** [comment]
@michaelbynum I'm moving this to the Aug release as the May release will be created next week.

**Comment by ksbeattie** [comment]
@michaelbynum any update on this?

**Comment by michaelbynum** [comment]
Apologies. I'll be pushing updates to address reviews early next week.

**Comment by ksbeattie** [comment]
> Apologies. I'll be pushing updates to address reviews early next week.

@michaelbynum I know you're busy... any news on this?

**Comment by ksbeattie** [comment]
@michaelbynum, I'm taking this off the Aug release board as it doesn't appear work is progressing here.
